File: 337-House Robber III.py
# 
# Runtime: 52 ms, faster than 99.74% of Python3 online submissions for House Robber III.
# Memory Usage: 14.5 MB, less than 96.61% of Python3 online submissions for House Robber III.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rob(self, root: 'TreeNode') -> 'int':
        return self.dfs(root)[1]

    def dfs(self, node):
        if node is None: return (0, 0)
        left, right = self.dfs(node.left), self.dfs(node.right)
        return (left[1] + right[1], max(left[1] + right[1], left[0] + right[0] + node.val))


'''
max(left[1] + right[1], left[0] + right[0] + node.val) 算是状态转移方程
    left[1] + right[1] 同一个父节点的两个节点
    left[0] + right[0] + node.val 两个孙子和一个爷爷
'''

File: Algorithms/C++/337-House_Robber_III.cpp
// Runtime: 8 ms, faster than 99.21% of C++ online submissions for House Robber III.
// Memory Usage: 20.5 MB, less than 100.00% of C++ online submissions for House Robber III.
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        return dfs(root).second;
    }
private:
    pair<int, int> dfs(TreeNode* node) {
        if (!node)
            return make_pair(0, 0);
        pair<int, int> l = dfs(node->left);
        pair<int, int> r = dfs(node->right);
        int f2 = l.second + r.second;                    // subtree max val
        int f1 = max(f2, l.first + r.first + node->val); // current node max val
        return make_pair(f2, f1);
    }
};
